drivers/modem/modem_cmd_handler.c

```diff
@@ -0,0 +1,488 @@
+			pos++;
+		}
+	}
+
+	if (buf && is_crlf(*(buf->data + pos))) {
+		len += pos;
+		*offset = pos;
+		*frag = buf;
+		return len;
+	}
+
+	return 0;
+}
+
+/*
+ * Cmd Handler Functions
+ */
+
+static inline struct net_buf *read_rx_allocator(s32_t timeout, void *user_data)
+{
+	return net_buf_alloc((struct net_buf_pool *)user_data, timeout);
+}
+
+/* return scanned length for params */
+static int parse_params(u8_t *buf, size_t buf_len, struct modem_cmd *cmd,
+			u8_t **argv, size_t argv_len, u16_t *argc)
+{
+	int i;
+	size_t begin = 0U, end = 0U;
+
+	if (!buf || !buf_len || !cmd || !argv || !argc) {
+		return -EINVAL;
+	}
+
+	while (end < buf_len) {
+		for (i = 0; i < strlen(cmd->delim); i++) {
+			if (buf[end] == cmd->delim[i]) {
+				/* mark a parameter beginning */
+				argv[*argc] = &buf[begin];
+				/* end parameter with NUL char */
+				buf[end] = '\0';
+				/* bump begin */
+				begin = end + 1;
+				(*argc)++;
+				break;
+			}
+		}
+
+		if (*argc == argv_len) {
+			break;
+		}
+
+		end++;
+	}
+
+	/* consider the ending portion a param if end > begin */
+	if (end > begin) {
+		/* mark a parameter beginning */
+		argv[*argc] = &buf[begin];
+		/* end parameter with NUL char
+		 * NOTE: if this is at the end of buf_len will probably
+		 * be overwriting a NUL that's already ther
+		 */
+		buf[end] = '\0';
+		(*argc)++;
+	}
+
+	/* missing arguments */
+	if (*argc < cmd->arg_count) {
+		return -EINVAL;
+	}
+
+	/*
+	 * return the beginning of the next unfinished param so we don't
+	 * "skip" any data that could be parsed later.
+	 */
+	return begin;
+}
+
+static void cmd_handler_process(struct modem_cmd_handler *cmd_handler,
+				struct modem_iface *iface)
+{
+	struct modem_cmd_handler_data *data;
+	struct modem_cmd *cmd;
+	struct net_buf *frag = NULL;
+	size_t out_len, rx_len, bytes_read = 0;
+	int i, j, ret;
+	u8_t *argv[CONFIG_MODEM_CMD_HANDLER_MAX_PARAM_COUNT];
+	u16_t offset, len, argc;
+	bool quit_loop;
+
+	if (!cmd_handler || !cmd_handler->cmd_handler_data ||
+	    !iface || !iface->read) {
+		return;
+	}
+
+	data = (struct modem_cmd_handler_data *)(cmd_handler->cmd_handler_data);
+
+	/* read all of the data from modem iface */
+	while (true) {
+		ret = iface->read(iface, data->read_buf,
+				  data->read_buf_len, &bytes_read);
+		if (ret < 0 || bytes_read == 0) {
+			/* modem context buffer is empty */
+			break;
+		}
+
+		/* make sure we have storage */
+		if (!data->rx_buf) {
+			data->rx_buf = net_buf_alloc(data->buf_pool,
+						     data->alloc_timeout);
+			if (!data->rx_buf) {
+				LOG_ERR("Can't allocate RX data! "
+					"Skipping data!");
+				break;
+			}
+		}
+
+		rx_len = net_buf_append_bytes(data->rx_buf, bytes_read,
+					      data->read_buf,
+					      data->alloc_timeout,
+					      read_rx_allocator,
+					      data->buf_pool);
+		if (rx_len < bytes_read) {
+			LOG_ERR("Data was lost! read %zu of %zu!",
+				rx_len, bytes_read);
+		}
+	}
+
+	/* process all of the data in the net_buf */
+	while (data->rx_buf) {
```

Comment:
the while loop below is hard to digest. way too long, would benefit to get parts split into inline functions.

Reply:
I can break this up a bit.